Implementing shlex module

rchinmay opened the pull request:
Implemented the Shlex module as a draft package module in `test/library/draft`, which will resolve the issue #11489, once promoted from a draft package to a package module in Chapel.

@ben-albrecht @bradcray @rapiz1 Please review this PR.

rapiz1 commented:
Good job! Finally someone gets to work on this.

But `test/library/packages/` is not the right place for the module, which as the name shows, contains tests. Placing it in `standard` or `package` or a separate mason package can be discussed by the core team.

Also, if you want to get this into the main branch instead of a separate package, you may want to align your code style to the chapel source. I don't know if there's a style guide. But at least keep consistent for spaces :)

rchinmay commented on test/library/packages/Shlex/Shlex.chpl:
```diff
@@ -0,0 +1,390 @@
+            if this.debug>=2{
+              writeln("shlex: I see EOF in quoted state");
+            }
+            throw new Error("No closing quotation");
```
@rapiz1 Thank you for taking a look at the PR. I believed the ``test/library/packages/`` path is for modules that have which are under testing and I thought it was a better place than ``modules/packages`` or ``modules/standard``. I will change it in the next commit. Also, I found that the tests were in different styles (like a different ``.chpl`` and ``.good`` folder for each category). I will update that as well in the next commit. I also wanted some other suggestions by which run-time errors could be generated in a user-friendly manner compared to ``throw new Error()`` format. 
I have followed the 80-column rule. Can you please elaborate on spacings as to where I might be wrong?

ben-albrecht replied:
I don't want to nitpick this PR to death, but I'll call out a few examples within, and you can apply them elsewhere. Things like this really make me wish we had an official style guide and linter...

bradcray commented:
> But test/library/packages/ is not the right place for the module, which as the name shows, contains tests.

It's OK to put draft modules under `test`, but `test/library/packages/` is intended to test modules that are already in `modules/packages`.  `test/library/draft/` would be a place to park a module (and its tests) that is not yet ready for `modules/packages`.  I don't have an immediate opinion as to whether this should live in `modules/packages` or be a mason package or something else and hope that @ben-albrecht will comment since he filed the original request.

ben-albrecht commented:
> I don't have an immediate opinion as to whether this should live in modules/packages or be a mason package or something else and hope that @ben-albrecht will comment since he filed the original request.

My preference is for this to initially be a package module (adding it as a draft package for this PR seems reasonable), which should eventually become a mason package when we solve the "package modules as mason packages" puzzle. The rationale for a package module is so that we can use this in mason to resolve a few issues encountered in parsing user inputs (https://github.com/chapel-lang/chapel/issues/15152).

@rchinmay - I plan on reviewing, thanks for your contribution and patience on us getting to this!

ben-albrecht commented on test/library/packages/Shlex/Test.chpl:
```diff
@@ -0,0 +1,167 @@
+use Shlex;
```
Are you familiar with Chapel's [testing system](https://chapel-lang.org/docs/master/developer/bestPractices/TestSystem.html#readme-testsystem)? Basically, you'll want to add a `Test.good` file that contains the expected output from this file. Also, maybe we can name these something like `ShlexTests.chpl` and `ShlexTests.good`? 

You can optionally split these tests up across multiple files for finer granularity testing as well as the ability to convert tests for features that not yet implemented into futures.

rchinmay commented:
@ben-albrecht I have updated the tests according to Chapel's testing system. Also, I have added a few new tests for the split function. I wanted to ask if there is any way of assigning ``NULL`` or ``c_nil`` to strings in Chapel, so that empty strings are not considered as EOF when ``posix`` is true. This would solve the bug of parsing empty strings with ``posix`` set to true. If it is not possible, I was thinking of using ``CPtr`` module and an additional pointer variable that refers to the token when it has not reached EOF, and stores ``c_nil`` when EOF is reached when ``posix`` is true. But, I believe it would complicate the design.

ben-albrecht commented:
> I have updated the tests according to Chapel's testing system.

Thanks @rchinmay - these are looking better.

> I wanted to ask if there is any way of assigning NULL or c_nil to strings in Chapel, so that empty strings are not considered as EOF when posix is true. This would solve the bug of parsing empty strings with posix set to true. If it is not possible, I was thinking of using CPtr module and an additional pointer variable that refers to the token when it has not reached EOF, and stores c_nil when EOF is reached when posix is true. But, I believe it would complicate the design.

I took a look at your implementation and the [Python implementation](https://github.com/python/cpython/blob/3.7/Lib/shlex.py). It's possible to make `this.eof` a generic field so that it can store the value `none` or a string, but only if you make the `posix` field a param (compile-time constant). Here's an example:

```
class C {
  var eof;

  proc init(param posix) {
    if posix {
      this.eof = none;
    } else {
      this.eof = '';
    }
  }
}
```

Otherwise, couldn't we check against the value of `posix` before checking the value of `eof` instead of using an `iseofnone` value? (Correct me if I'm wrong, I haven't reviewed the full implementation in detail yet)

rchinmay commented:
@ben-albrecht 
> It's possible to make this.eof a generic field so that it can store the value none (type nothing) or a string

I tried doing this, but in the ``get_token()`` function, the condition checking for eof would be modified to
```chpl
if(this.tokindex==-1){
  return this.eof;
}
```
And I was getting an error
```
error: unable to resolve return type
```
So, I decided to go with the 2nd idea of testing if the token is an EOF, from ``tokindex`` ( which points to index in the string that we are reading), and not from ``token`` when ``posix`` is set to ``true``. This should not cause any incorrect results, because I am assuming no string can contain the EOF character before the string ends. But I am afraid if there was a neater implementation. However, the tests with empty strings also pass and match with the output from the python implementation. If you think that the implementation should have been different, please inform me.
Thanks.

rchinmay commented:
@ben-albrecht Tests which I tried manually which had trailing whitespaces were failing due to the change in procedure for checking ``EOF``. For example, the string ``abc `` (with trailing space) passed into split function would give result as ``['abc','']`` since the trailing space returned empty string which is not ``EOF`` but was inserted to list as empty string, which is wrong. So I tried stripping the trailing spaces in the init function itself, which solved the problem.

ben-albrecht commented:
Thanks for the update @rchinmay. I started reviewing this in more depth today and anticipate finalizing and posting my review comments tomorrow. Thanks for your patience!

ben-albrecht commented on test/library/packages/Shlex/Shlex.chpl:
```diff
@@ -0,0 +1,385 @@
+        else if (this.quotes.find(this.state)!=-1){
+          quoted=true;
+          if nextchar==''{
+            if this.debug>=2{
```
formatting nit: I think the spaces here are generally more readable (and I think this is what @rapiz1 was suggesting):

```suggestion
            if this.debug >= 2 {
```

ben-albrecht commented on test/library/packages/Shlex/Shlex.chpl:
```diff
@@ -0,0 +1,385 @@
+          if (nextchar==this.state){
+            if !this.posix{
+              this.token+=nextchar;
+              this.state=' ';
```
formatting nit:

```suggestion
              this.state = ' ';
```

ben-albrecht commented on test/library/packages/Shlex/Shlex.chpl:
```diff
@@ -0,0 +1,385 @@
+
+  //
+  // This is the public interface to the SHLEX.
+  // Concatenates the tokens of the list passed into a string.
```
Check out [chpldoc](https://chapel-lang.org/docs/master/tools/chpldoc/chpldoc.html) if you haven't already. We can generate documentation from the module if you use the format:

```
/*

documentation

*/
proc join(...) ... {
   ...
}

ben-albrecht commented on test/library/packages/Shlex/Shlex.chpl:
```diff
@@ -0,0 +1,385 @@
+  }
+
+  //
+  // This is the public interface to the SHLEX.
```
I think we can get rid of this line about public interface

ben-albrecht commented on test/library/packages/Shlex/Shlex.chpl:
```diff
@@ -0,0 +1,385 @@
+      compilerError("join() only takes homogeneous tuple of strings " +
+                      "or array of strings");
+    if x.size == 1 {
+      if isArray(x) then return x[x.domain.first];
```
Does this work?

```suggestion
      if isArray(x) then return x.first;
```

ben-albrecht commented on test/library/packages/Shlex/Shlex.chpl:
```diff
@@ -0,0 +1,385 @@
+    var res="";
+    var beginning: bool = false;
+    for ele in x do {
+      if(beginning) then res+=" ";
```
formatting nit:

```suggestion
      if beginning then res += " ";
```

ben-albrecht reviewed (commented):
@rchinmay - Thank you for this excellent contribution! This will be valuable for mason and future command line tools written in Chapel. 

The implementation looks good overall. Most of my feedback involves smaller changes, layout suggestions, or formatting tips (I only gave a few sample suggestions, which you can extrapolate to the rest of your code).

Can you put this module into `test/library/draft/Shlex/` and all tests into `test/library/draft/Shlex/API` ?

rchinmay commented:
@ben-albrecht Since I had not worked with ``chpldoc`` earlier, I had some issues regarding its output. The ``split`` function has its argument list messed up in ``chpldoc`` generated web-view. It is definitely my mistake, due which this problem has been caused. I tried adding indentation, but it gave errors as 
``Warning, treated as error:
/tmp/chpldoc-rchinmay.deleteme-hQFolm/source/modules/Shlex.rst:124:Unexpected indentation.``
Also, the line number does not match with any comment. Also, the log file would be deleted after showing the error. So it was a difficult time debugging for a first-time ``chpldoc`` user.  By removing the indentation, the output of web-view was messed up. I will solve the problem by tomorrow, by looking at other modules. 
I have pushed the changes because I think I have completed the styling suggestions as you advised. Please inform if any other styling changes is needed. Also, I tried changing the path, but I could not figure out how to use modules from a different path. I tried ``use ../Shlex`` but it did not work. Can you please tell me how to use or import module's functions from a file located in different directory?
Thanks.

ben-albrecht commented:
> Also, the line number does not match with any comment. Also, the log file would be deleted after showing the error. So it was a difficult time debugging for a first-time chpldoc user.

I think that's expected currently. chpldoc generates an rst file which is then generated into the HTML. The error is pointing to the generated rst file. I encourage you to file an issue about this (if we don't already have one) to track the feature request of improving this user experience.

> I have pushed the changes because I think I have completed the styling suggestions as you advised. Please inform if any other styling changes is needed. 

Thanks, taking a look...

> Also, I tried changing the path, but I could not figure out how to use modules from a different path. I tried use ../Shlex but it did not work. Can you please tell me how to use or import module's functions from a file located in different directory?

Add a `COMPOPTS` file to the `Shlex/API/` subdirectory with the contents:

```
-M ../
```

This will add this flag to all compilations with `start_test` in this directory. `-M <path>` is how you add to the module search path.

ben-albrecht commented on test/library/packages/Shlex/Shlex.chpl:
```diff
@@ -0,0 +1,442 @@
+module Shlex {
+  private use List;
+  private use Regex;
+  class shlex{
```
```suggestion
  class shlex {
```

ben-albrecht replied:
Note: I _think_ committing these suggestions will generate an error in the DCO signed author checks, so I believe it's better for you to make the changes yourself in the code.

ben-albrecht commented on test/library/packages/Shlex/Shlex.chpl:
```diff
@@ -0,0 +1,442 @@
+      :rtype: `string`
+      :throws Error: Unbalanced quotes or absence of escaped character.
+    */
+    proc read_token(): string throws{
```
```suggestion
    proc read_token(): string throws {
```

ben-albrecht replied:
Note: I _think_ committing these suggestions will generate an error in the DCO signed author checks, so I believe it's better for you to make the changes yourself in the code.

ben-albrecht commented on test/library/packages/Shlex/Shlex.chpl:
```diff
@@ -0,0 +1,442 @@
+      :throws Error: Unbalanced quotes or absence of escaped character.
+    */
+    proc read_token(): string throws{
+      var quoted:bool = false;
```
no change needed, but as a tip:

You don't need to declare the type when it is being initialized like this. It could instead be:

```chpl
var quoted = false;
```

ben-albrecht commented on test/library/packages/Shlex/Shlex.chpl:
```diff
@@ -0,0 +1,442 @@
+          nextchar = this.instream[this.tokindex];
+          this.tokindex += 1;
+        }
+        else{
```
```suggestion
        else {
```

ben-albrecht replied:
Note: I _think_ committing these suggestions will generate an error in the DCO signed author checks, so I believe it's better for you to make the changes yourself in the code.

ben-albrecht commented on test/library/packages/Shlex/Shlex.chpl:
```diff
@@ -0,0 +1,442 @@
+module Shlex {
+  private use List;
+  private use Regex;
+  class shlex{
```
I think we can `pragma "no doc"` this class, since it isn't user-facing

ben-albrecht commented on test/library/packages/Shlex/Shlex.chpl:
```diff
@@ -0,0 +1,442 @@
+                                      set to true. Defaults to `();|&`.
+      :type custom_punctuation_chars: `string`
+    */
+    proc init(instream:string, posix: bool = false,
```
I have a feeling your chpldoc error is coming from somewhere in the docstring above. Does this initializer need to be publicly documented? Will users ever initialize their own shlex class? I assumed they would use the standalone functions.

rchinmay replied:
Some of the conditions can not be initialized by ``split`` function, such as having ``punctuation_chars`` set to true and ``whitespace_split`` set to true simultaneously, or both of them to false. One of the tests present in ``ParsingEmptyStrings.chpl`` have such a requirement where ``split`` function does not satisfy. Also, other requirements such as changing ``debug`` or ``wordchars`` could be done easily if an user is allowed to use ``init`` function. If we have to make the standalone function ``split()`` to handle these requirements, its argument list would also be same as the variables in ``shlex`` class. Please confirm if ``init`` function should not be user-facing, or is it better to document it and allow the users to configure necessary variables. Also, the Python shlex implementation has less user flexibility provided by ``split`` function, and most of its test uses constructor for configuring  variables.

ben-albrecht reviewed (commented):
Few more comments. Formatting looks a lot better!

rchinmay commented:
@ben-albrecht I have corrected the comments so that their web view is as expected. I had another query related to ``chpldoc``. Can you please confirm if it is possible for a class to have ``pragma "no doc"`` and some of its functions to have documentation? I wanted the ``shlex`` class not to have its list of variables included in the documentation. But I wanted it's ``init`` and ``get_token`` function to have documentation. But if I insert ``pragma "no doc"`` above the class, then the documentation of ``init`` function and other functions are also not displayed in the documentation. Is there any workaround for this issue, other than having a standalone ``init`` function, with a different name, that the users could access and its documentation is visible?
Thanks.

ben-albrecht commented:
@rchinmay - I mistakenly assumed we did not want to expose the class as part of the user API (since I have never used the class interface in Python's shlex module before). I suggest that you do not no-doc the `Shlex` class, and instead no-doc any fields or methods you do not want to expose to the user interface.

> Can you please confirm if it is possible for a class to have pragma "no doc" and some of its functions to have documentation?

I don't think we want to do this anymore, but this is possible if you were to make the methods tertiary methods, which look like this:

```chpl
pragma "no doc"
class C {
  proc init() { }
}

/* Doc string */
proc C.someMethod() {

}
```

ben-albrecht commented on test/library/draft/Shlex/Shlex.chpl:
```diff
@@ -0,0 +1,443 @@
+/*
```
Can you add an empty `Shlex.notest` file to this directory so that `start_test` doesn't assume this file is a test?

rchinmay commented:
@ben-albrecht I have made changes in documentation and some styling changes in test files. Please review the changes. 
Thanks.

ben-albrecht reviewed (approved):
Looks great, thanks @rchinmay!

ben-albrecht commented:
@rchinmay - I modified the PR description and removed this section which is no longer true (let me know if you disagree with any modifications):

> Please note that a bug exists when parsing empty strings with posix set to true. All other tests in [python test library](https://github.com/python/cpython/blob/3.7/Lib/test/test_shlex.py), and other tests that I tried manually are working well.
Since the EOF is to be set as None (or nil) and I could not find how to assign strings to None (or nil), I decided to go further with keeping it as empty strings, but in tests, I found that parsing empty strings with posix set to true, gives incorrect output, as it thinks that it is equivalent to EOF. I thought that we could have another variable to check if the token is None, but I wanted to confirm this as it was a design issue.
